### PR TITLE
Deploy and enable the UI plugin by nmstate-operator

### PR DESCRIPTION
**Pull request by orenc1**
Upon NMState CR creation, when running on OpenShift, the nmstate operator controller will create and reconcile the following resources required to deploy the UI plugin in OCP console:
* plugin Deployment (nginx server serving the UI artifacts)
* plugin Service (to serve the UI contents in the cluster internally)
* ConfigMap (nginx configuration)
* ConsolePlugin Custom Resource
* Add the console plugin to Console.operator.openshift.io plugins list

Signed-off-by: Oren Cohen <ocohen@redhat.com>

**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug

/kind enhancement

**What this PR does / why we need it**:

**Special notes for your reviewer**:

**Release note**:

```release-note
Deploy NMState UI plugin on OpenShift by nmstate-operator upon installation.
```

### CODE REVIEW

**Comment by orenc1** [comment]
@yzamir, @upalatucci - this is the PR that will deploy and activate the nmstate console plugin on OpenShift upon creating the NMState custom resource.

@phoracek / @qinqon / @RamLavi - could you please take a look?

File: deploy/operator/role.yaml [hunk 1]
```diff
@@ -53,12 +53,27 @@ rules:
...
   - '*'
+- apiGroups:
+  - operator.openshift.io
+  resources:
+  - consoles
+  verbs:
+  - get
+  - list
+  - update
+  - watch
```
Comment (qinqon) [inline deploy/operator/role.yaml:76]: This things has to be present only if openshift is detected, we don't want to polute non openshift deploys.

File: deploy/ui-plugin/configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+apiVersion: v1
```
Comment (qinqon) [inline deploy/ui-plugin/configmap.yaml:1]: can we move all this `deploy/openshift/ui-plugin` ? at the end this only works for openshift

File: controllers/operator/nmstate_controller.go [hunk 3]
```diff
@@ -389,3 +439,19 @@ func setClusterReaderExist(c client.Client, data render.RenderData) error {
 	data.Data["ClusterReaderExists"] = found
 	return nil
 }
+
+func getEnv(key, fallback string) string {
```
Comment (qinqon) [inline controllers/operator/nmstate_controller.go:443]: There is a `pkg/environment/environment` module for this, can you move there and rename to follow the pattern ?

File: controllers/operator/nmstate_controller.go [hunk 4]
```diff
@@ -44,6 +44,7 @@ import (
 	nmstatev1 "github.com/nmstate/kubernetes-nmstate/api/v1"
 	"github.com/nmstate/kubernetes-nmstate/pkg/cluster"
 	nmstaterenderer "github.com/nmstate/kubernetes-nmstate/pkg/render"
+	operatorv1 "github.com/openshift/api/operator/v1"
```
Comment (qinqon) [inline controllers/operator/nmstate_controller.go:47]: The context of this code is not `openshift`
```suggestion
	openshiftoperatorv1 "github.com/openshift/api/operator/v1"
```

File: controllers/operator/nmstate_controller.go [hunk 5]
```diff
@@ -289,12 +306,45 @@ func (r *NMStateReconciler) applyHandler(instance *nmstatev1.NMState) error {
 	return r.renderAndApply(instance, data, "handler", true)
 }
 
+func (r *NMStateReconciler) applyUIPlugin(instance *nmstatev1.NMState) error {
```
Comment (qinqon) [inline controllers/operator/nmstate_controller.go:309]:
```suggestion
func (r *NMStateReconciler) applyOpenshiftUIPlugin(instance *nmstatev1.NMState) error {
```

File: controllers/operator/nmstate_controller.go [hunk 6]
```diff
@@ -289,12 +306,45 @@ func (r *NMStateReconciler) applyHandler(instance *nmstatev1.NMState) error {
...
+	data := render.MakeRenderData()
+	data.Data["PluginNamespace"] = getEnv("HANDLER_NAMESPACE", "openshift-nmstate")
+	data.Data["PluginName"] = getEnv("PLUGIN_NAME", "nmstate-console-plugin")
+	data.Data["PluginImage"] = getEnv("PLUGIN_IMAGE", "quay.io/nmstate/nmstate-console-plugin:release-1.0.0")
+	return r.renderAndApply(instance, data, "ui-plugin", true)
+}
+
+func (r *NMStateReconciler) patchConsolePlugin(ctx context.Context) error {
```
Comment (qinqon) [inline controllers/operator/nmstate_controller.go:317]:
```suggestion
func (r *NMStateReconciler) patchOpenshiftConsolePlugin(ctx context.Context) error {
```

**Review by qinqon — CHANGES_REQUESTED** [review]
Let's move manifests to deploy/openshift and also add openshift context to functions and import also some unit test is missing since we are not able to do e2e openshift test at this repo.

File: deploy/operator/role.yaml (hunk 1 above)
Comment (orenc1) [reply deploy/operator/role.yaml:76]: We're using kubebuilder markers (which are comments) to generate the RBAC roles.
How could it be possible to configure conditionally this way?

File: controllers/operator/nmstate_controller.go [hunk 7]
```diff
@@ -289,12 +306,45 @@ func (r *NMStateReconciler) applyHandler(instance *nmstatev1.NMState) error {
...
+	}
+
+	if !stringInSlice(pluginName, consoleObj.Spec.Plugins) {
+		r.Log.Info("Enabling kubevirt plugin in Console")
+		consoleObj.Spec.Plugins = append(consoleObj.Spec.Plugins, pluginName)
+		err := r.Client.Update(ctx, consoleObj)
+		if err != nil {
+			r.Log.Error(err, fmt.Sprintf("Could not update resource - APIVersion: %s, Kind: %s, Name: %s",
+				consoleObj.APIVersion, consoleObj.Kind, consoleObj.Name))
+			return err
```
[the remark this answers is not in the document]
Comment (orenc1) [reply controllers/operator/nmstate_controller.go:335]: not sure about that.
isn't the reconciler looping again on errors anyway until current config == desired config?

File: controllers/operator/nmstate_controller.go (hunk 3 above)
Comment (orenc1) [reply controllers/operator/nmstate_controller.go:443]: moved that function to the `environment` package

File: controllers/operator/nmstate_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply controllers/operator/nmstate_controller.go:334]: Yep, so we good.

File: deploy/operator/role.yaml (hunk 1 above)
Comment (qinqon) [reply deploy/operator/role.yaml:76]: Nope, let's just leave like that

File: controllers/operator/nmstate_controller.go (hunk 4 above)
Comment (orenc1) [reply controllers/operator/nmstate_controller.go]: also changed: `consolev1` --> `openshiftconsolev1`

**Pull request merged**